File: Ejemplos/Ficheros-Java/src/main/java/es/joseluisgs/serializable/Serializable.java
package es.joseluisgs.serializable;


import java.io.*;
import java.util.ArrayList;
import java.util.List;

public class Serializable {
    public static void main(String[] args) {
        System.out.println("Hola Serializable");

        escribirSerializable();
        leerSerializable();
    }

    private static void leerSerializable() {
        String path = System.getProperty("user.dir") + File.separator + "data";
        String personasFile = path + File.separator + "personas.dat";

        var personas = new ArrayList<Persona>();

        try (var ois = new ObjectInputStream(new FileInputStream(personasFile))) {
            // Podemos leer un objeto o coleccion de objetos
            personas = (ArrayList<Persona>) ois.readObject();
            // O poder leerlo de esta forma
            /*while (ois.available() > 0) {
                personas.add((Persona) ois.readObject());
            }*/
        } catch (IOException | ClassNotFoundException e) {
            e.printStackTrace();
        }
        System.out.println("Personas: " + personas.size());
        personas.forEach(System.out::println);
    }

    private static void escribirSerializable() {
        String path = System.getProperty("user.dir") + File.separator + "data";
        String personasFile = path + File.separator + "personas.dat";

        ArrayList<Persona> personas = new ArrayList<>(
                List.of(
                        new Persona("Pepe", 20),
                        new Persona("Juan", 30),
                        new Persona("Luis", 40)
                )
        );

        try (var oos = new ObjectOutputStream(new FileOutputStream(personasFile))) {
            // Podemos escribir un objeto o coleccion de objetos
            oos.writeObject(personas);
            // personas.forEach(p -> oos.writeObject(p));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

// Para que un objeto sea serializable debe implementar la interfaz java.io.Serializable
// y todos sus atributos deben ser serializables
class Persona implements java.io.Serializable {
    int edad;
    private String nombre;

    public Persona(String nombre, int edad) {
        this.nombre = nombre;
        this.edad = edad;
    }

    public String getNombre() {
        return nombre;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public int getEdad() {
        return edad;
    }

    public void setEdad(int edad) {
        this.edad = edad;
    }

    @Override
    public String toString() {
        return "Persona{" + "nombre=" + nombre + ", edad=" + edad + '}';
    }

}

File: Ejemplos/Ficheros-Kotlin/src/main/kotlin/serializable/Serializable.kt
package serializable

import java.io.*


// Para ser serializable todas las propiedades deben ser serializables (sublases incluídas)
private data class Persona(val nombre: String, val edad: Int) : Serializable

fun main() {
    escribirFicheroSerializable()
    leerFicheroSerializable()
}

fun leerFicheroSerializable() {
    val programPath = System.getProperty("user.dir")
    val filePath = "$programPath${File.separator}data${File.separator}fichero-serializable.dat"
    println("Leyendo fichero: $filePath")

    val file = File(filePath)

    val input = ObjectInputStream(FileInputStream(file))
    input.use {
        // Podemos leer cualquier objeto serializable, ya sea un objeto o una colección
        val persona = it.readObject() as Persona
        val lista = it.readObject() as List<Persona>
        println(persona)
        println(lista)
    }
}

fun escribirFicheroSerializable() {
    val programPath = System.getProperty("user.dir")
    val filePath = "$programPath${File.separator}data${File.separator}fichero-serializable.dat"
    println("Escribiendo fichero: $filePath")

    val file = File(filePath)
    val persona = Persona("Antonio", 34)
    val lista = listOf(
        Persona("Pepe", 23),
        Persona("Juan", 34),
        Persona("Ana", 45)
    )
    val output = ObjectOutputStream(FileOutputStream(file))
    output.use {
        // Podemos escribir cualquier objeto serializable, ya sea un objeto o una colección
        it.writeObject(persona)
        it.writeObject(lista)
    }
}
